keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -12,124 +11,477 @@ approvers:
+in active use.
+
+We have a couple of choices (and non-choices, see appendix) for the
+current resource:
+
+1.  We can delete the current resource from extensions.v1beta1 in
+    anticipation that an improved API can replace it.
+
+1.  We can copy the API as-is (or with minor changes) into
+    networking.v1beta1, preserving/converting existing data (following
+    the same approach taken with all other extensions.v1beta1
+    resources). This will allow us to start the cleanup of the
+    extensions API group. This also prepares the API for GA.
+
+Option 1 does not seem realistic in a short-term time frame (a new API
+will need to be taken through design, alpha/beta/ga phases). At the
+same time, there are enough users that the existing API cannot be
+deleted out right.
+
+In terms of moving the API towards GA, the API itself has been
+available in beta for so long that it has attained defacto GA status
+through usage and adoption (both by users and by load balancer /
+ingress controller providers). Abandoning it without a full
+replacement is not a viable approach.  It is clearly a useful API and
+captures a non-trivial set of use cases.  At this point, it seems more
+prudent to declare the current API as something the community will
+support as a V1, codifying its status, while working on either a V2
+Ingress API or an entirely different API with a superset of features.
 
 ### Goals
 
-* Move Ingress to a permanent API group.
-* Position the Ingress API in a place towards making progress towards GA.
-
-## Proposal
+A detailed list of the changes being proposed is given in the Design
+section below.
+
+* Move Ingress to a permanent API group. (status: implemented)
+* Make changes to the Ingress API be in a GA-ready state. (status:
+  proposal).
+  * Clean up the Ingress API (fix ambiguities, API spec bugs).
+  * Promote commonly supported annotations to proper API fields.
+  * Create a suite of conformance tests to validate existing
+    implementations.
+* Make Ingress GA. (status: proposal).
+
+## Design
+
+This section describes the API fixes proposed for GA.
+
+### Summary of the proposed changes
+
+1. Add path as a prefix and make regex support optional. The current spec states
+   that the path is a regular expression and arbitrary regular expression, but
+   support varies across providers. In addition, regex matching is not supported
+   by many of the cloud provider implementations.
+1. Fix API field naming:
+   1. `spec.backend` should be called `spec.defaultBackend`.
+1. Hostname wildcard matching. We currently allow for creation of
+   `*.foo.com` and this seems to be a commonly suppported host match,
+   but this is not part of the spec.
+1. Specify healthcheck behavior and be able to configure the healthcheck path
+   and timeout.
+1. Improve the Ingress status field to be able to include additional
+   information. The current status currently only contains provisions the IP
```

Comment:
did you mean `provisioned IP`?

Reply:
yes ,thanks

Reply:
Hi, shouldn't the health check for service backend be part of ServiceSpec instead of Ingress ?
Also, given we have pod readness probe, and endpoint controller, and Node's ready status, do we really need healthChecks from cloudProvider's Loadbalancer? (there will be two duplicated healthcheck mechanism to pods, one from cloudProvider and another from pod readness  prod)

Reply:
> Hi, shouldn't the health check for service backend be part of ServiceSpec instead of Ingress ?

While yes, this is true, it's a little weird to describe an L7 healthcheck in an otherwise L4 concept.  I'm not sure if it would be better.

> given we have pod readness probe, and endpoint controller, and Node's ready status, do we really need healthChecks from cloudProvider's Loadbalancer?

pod readiness is very localized. It seems that most cloud providers want something at their API to be doing health monitoring, so I think we need to offer some way to express this.  Th eheuristic we use today is just bad.